### ISSUE TITLE
[RFC] Adding the Pimple dumper as a require-dev

### ISSUE THREAD

**Issue by GawainLynch** [bolt/bolt#6394]
## Background

As documented in the [`bolt/bolt` wiki](https://github.com/bolt/bolt/wiki/Developing-Bolt-in-PHPStorm) there is a Composer package [`sorien/silex-pimple-dumper`](https://github.com/Sorien/silex-pimple-dumper) that can be used to build a JSON file with most of the Pimple services used in Bolt.

This is used for two purposes,  firstly to allow hinting and code completion in PhpStorm (and in future possibly other IDEs), and is used to generate the [Bolt Class Reference](https://docs.bolt.cm/class-reference) page in our documentation.

Currently the core team uses a combination of PhpStorm/IntelliJ, vim, Atom, Sublime, Netbeans, and Visual Studio Code. 

So please, in no way should this RFC be seen as trying to lock/force anyone into using JetBrain's IDEs, rather just integrating some useful stuff into stuff for stuff … 'n stuff.

## Problem

The current approach has `sorien/silex-pimple-dumper` version 1.x installed as a Composer global package, that is serving us fine for Bolt 3.x development as we're locked into Silex 1.

However, Silex 2 requires `sorien/silex-pimple-dumper` version 2 (due to changes in Pimple, etc), and if you spend part of your day working on Bolt 3.x, part on 4.x (`master`), and part on Silex v2 applications which are exploding in terms of numbers on my workstation alone, you spend a lot of time changing the globally installed `sorien/silex-pimple-dumper` version back and forward.

## Proposal

1. Add `sorien/silex-pimple-dumper` to the Bolt `"require-dev"`
2. Add `pimple.json` to the git repository's `.gitignore` file
3. Add the script to generate `pimple.json` to Bolt's git repository

This will only affect **git** installs of Bolt, as `require-dev` packages are only installed based on the *root package*, so archive and normal Composer installs of Bolt will never have the library appear in their `vendor/`.

**Comment by GawainLynch** [comment]
**Update:** The hinting & code completion I refer to above is for services on the application container.

e.g. `$app['request']->fooBar()` would display a warning, as `Symfony\Component\HttpFoundation\Request` does not have a method called `fooBar()`, however using your configured hotkey on `$app['request']->getClientIp()` would take you to the actual function code in the `Request` class.

This also allows you to do things like verify expected return types, etc … time & stress saving level Over 9,000!

**Comment by bobdenotter** [comment]
Fine with me! 👍

**Comment by GawainLynch** [comment]
I have created a [repository to house this](https://github.com/bolt/ide-tools) and created an RFC branch on [`bolt/bolt`](https://github.com/bolt/bolt/tree/dev-feature/squeeze-my-pimple) that can be checked out, and tested by running `./vendor/bin/pimple-dumper`

Feedback most welcome!

**Comment by CarsonF** [comment]
I'm in favor of dev requirement. Let's make it a nut command instead. 

Here's what GMO uses:
```php
<?php

namespace GMO\Console;

use Silex;
use Sorien\Provider\PimpleDumpProvider;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\HttpFoundation\Request;

class PimpleDumpCommand extends ContainerAwareCommand
{
    public function isEnabled()
    {
        return class_exists('\Sorien\Provider\PimpleDumpProvider');
    }

    protected function configure()
    {
        parent::configure();
        $this->setName('pimple:dump')
            ->setDescription('Dump the pimple container for the idea plugin')
        ;
    }

    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $app = $this->getContainer();
        if (!$app instanceof Silex\Application) {
            throw new \InvalidArgumentException('Container given needs to be an instance of Silex\Application');
        }

        $app['debug'] = true;

        $dumper = new PimpleDumpProvider();
        $app->register($dumper);
        $dumper->boot($app);

        $request = Request::create('/');
        $response = $app->handle($request);
        $app->terminate($request, $response);
    }
}
```

**Comment by GawainLynch** [comment]
On it

**Comment by GawainLynch** [comment]
Side note: That is what I wanted to do, but figured you'd naxx it :laughing:

**Comment by GawainLynch** [comment]
Addressed in #6402
